cpp/src/join/conditional_join.cu

```diff
@@ -69,8 +69,10 @@ conditional_join(table_view const& left,
   // If none of the input columns actually contain nulls, we can still use the
   // non-nullable version of the expression evaluation code path for
   // performance, so we capture that information as well.
-  auto const nullable  = cudf::nullable(left) || cudf::nullable(right);
-  auto const has_nulls = nullable && (cudf::has_nulls(left) || cudf::has_nulls(right));
+  auto const expr_has_nulls = ast::detail::contains_null_literal(&binary_predicate);
+  auto const nullable       = cudf::nullable(left) || cudf::nullable(right) || expr_has_nulls;
+  auto const has_nulls =
+    (nullable && (cudf::has_nulls(left) || cudf::has_nulls(right))) || expr_has_nulls;
```

Comment:
I suspected that both your suggestion and your question might come up during review :) The reason that I didn't implement the function that you propose is because expressions don't actually have access to the column values. An `ast::column_reference` object consists of an integer index and an `ast::table_reference`, but no corresponding reference to an actual `cudf::table_view` object. Practically we could simply redefine your proposed function as `contains_nulls(ast::expression&expr, cudf::table_view const& left, cudf::table_view const& right)`. However, it's important to recognize that this is also a _conceptually_ different question. An expression (generically, not just in the libcudf context) is composed of constants, variables, and operators/functions. Whether or not an expression intrinsically contains null _constant_ values is different from asking whether evaluating an expression could result in null output if we have some information about the values of the _variables_. I would be fine redefining this function to check the latter, we would just want to use a different name, perhaps something like `may_evaluate_null`.

If we redefined it this way, then I would be fine making it a method of `ast::expression`. The current version (being very literal-specific) didn't make sense as a class method.

Reply:
Oops, I meant to include the table references in my proposed API and forgot. Your redefinition is what I was imagining.

I think the advantage of the `may_evaluate_null` method is that you only have to check the nullability of things that are actually used. Currently if an expression for a computed column just returns a constant literal (`f(x) = 1`) but is evaluated on a table containing one or more nullable columns, then the code will needlessly use the nullable kernel template. Traversing the expression tree and checking the relevant literals/columns/operators would be more correct (no false positive in the constant case mentioned above) and more efficient (no need to check unused columns for nullability).

Reply:
Done.